### PR TITLE
Add support for generating the `.. x ? [y] : []` pattern in a collection-expression

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Part of: https://github.com/dotnet/roslyn/issues/69132

This change makes it so that if we see:

```c#
List<int> list = new List<int>();
list.Add(1);
if (x)
    list.Add(2);
```

We now generate: `[1, .. x ? [2] : []]`.  

This is a form which we view as idiomatic for inserting elements conditionally, and will be optimized by the compiler.

As i've been working in this space, i've been pulling out cleanups that i think help the code.  This PR is a collection of them.  

THis also adds a ton more tests validating behavior.

Note: next PR works a lot on improving the formatting of the produced collection-expressions.

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/CSharpAnalyzers.projitems [hunk 1]
```diff
@@ -83,6 +83,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\CSharpUseCollectionExpressionForEmptyDiagnosticAnalyzer.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\CSharpUseCollectionExpressionForArrayDiagnosticAnalyzer.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\UseCollectionExpressionHelpers.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)UseCollectionInitializer\CSharpUseCollectionInitializerAnalyzer.cs" />
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/CSharpAnalyzers.projitems:86]: there used to be a single `UseCollectionInitializerAnalyzer` type that was used for both VB and C#.  HOwever, it was getting crushed under the weight of having to have special logic (esp. logic that was really only for C#'s collection expressions).  

We now use inheritance here, which cleans up a few things (especially around type arguments).

File: src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        CSharpUseCollectionInitializerAnalyzer>
+{
+    protected override bool IsComplexElementInitializer(SyntaxNode expression)
+        => expression.IsKind(SyntaxKind.ComplexElementInitializerExpression);
+
+    protected override void GetPartsOfForeachStatement(
+        ForEachStatementSyntax statement,
+        out SyntaxToken identifier,
+        out ExpressionSyntax expression,
+        out IEnumerable<StatementSyntax> statements)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerAnalyzer.cs:31]: this is a better way than how i was previously doing things iwth ISYntaxFacts.  The problem is that there are too many foreach-forms across C# and VB (and in C# itself).  So syntaxfacts didn't really know how to deal with things.

Here though we know exactly what form we have, and we can produce a good output result for that for C# no matter what.

File: src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        identifier = statement.Identifier;
+        expression = statement.Expression;
+        statements = ExtractEmbeddedStatements(statement.Statement);
+    }
+
+    protected override void GetPartsOfIfStatement(
+        IfStatementSyntax statement,
+        out ExpressionSyntax condition,
+        out IEnumerable<StatementSyntax> whenTrueStatements,
+        out IEnumerable<StatementSyntax>? whenFalseStatements)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionInitializer/CSharpUseCollectionInitializerAnalyzer.cs:42]: same issue.  C# has `if/else` (Where else is optional).  VB has `if/elif/elif/else`.  These are really different forms, and trying to shoehorn into ISyntaxFacts was not the right way to do things.

File: src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider.cs [hunk 4]
```diff
@@ -140,42 +148,48 @@ private static ExpressionSyntax ConvertInvocation(InvocationExpressionSyntax inv
...
-                if (expression is InitializerExpressionSyntax or AssignmentExpressionSyntax)
-                    return true;
-
-                // if any of the expressions we're adding are multiline, then make things multiline.
-                if (!sourceText.AreOnSameLine(expression.GetFirstToken(), expression.GetLastToken()))
-                    return true;
-            }
-
             var totalLength = "{}".Length;
             foreach (var match in matches)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider.cs:152]: note: while this is getting cleaned up here, it is also getting refactored heavily in the next PR where i redo how we do formatting here entirely.  so don't have to examien this *too* closely :)

File: src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests.cs [hunk 5]
```diff
@@ -154,6 +154,122 @@ void M(int[] x)
...
+                class C
+                {
+                    void M(bool b)
+                    {
+                        var c = new List<int>
+                        {
+                            1
+                        };
+                        if (b)
+                            c.Add(2);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests.cs:187]: these tests validate that for the *collection INITIALIZER* space we don't do anything special.  The new form is for the *collection EXPRESSION* case.

File: src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests_CollectionExpression.cs [hunk 6]
```diff
@@ -97,6 +97,304 @@ void M()
...
+            }
+            """,
+            """
+            using System.Collections.Generic;
+
+            class C
+            {
+                void M(bool b)
+                {
+                    List<int> c = [1, .. {|CS0173:b ? [2] : []|}];
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests_CollectionExpression.cs:125]: this error is going to be fixed by chuck soon.

File: src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests_CollectionExpression.cs [hunk 7]
```diff
@@ -1802,4 +2137,164 @@ public async Task TestInTopLevelStatements()
...
+            """,
+            """
+            using System.Collections.Generic;
+
+            class C
+            {
+                void M()
+                {
+                    List<int> c = [1
+            ];
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/UseCollectionInitializer/UseCollectionInitializerTests_CollectionExpression.cs:2195]: this formatting is terrible.  next PR fixes this up.  but it's a lot of manual trivia handling, so i'm pulling it out of this PR to keep things sane.

File: src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs [hunk 8]
```diff
@@ -13,53 +14,88 @@
...
+            TExpressionSyntax,
+            TStatementSyntax,
+            TObjectCreationExpressionSyntax,
+            TMemberAccessExpressionSyntax,
+            TInvocationExpressionSyntax,
+            TExpressionStatementSyntax,
+            TForeachStatementSyntax,
+            TIfStatementSyntax,
+            TVariableDeclaratorSyntax,
+            TAnalyzer>, new()
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs:52]: yes.  this part is brain melting.  but it means the *consumption* side of this type (which references the derived types) now gets *much simpler*.  WIll point that out when we get tehre.

File: src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs [hunk 9]
```diff
@@ -81,13 +117,19 @@ protected override void AddMatches(ArrayBuilder<Match<TStatementSyntax>> matches
             if (initializer != null)
             {
                 var firstInit = _syntaxFacts.GetExpressionsOfObjectCollectionInitializer(initializer).First();
+
+                // if we have an object creation, and it *already* has an initializer in it (like `new T { { x, y } }`)
+                // this can't legally become a collection expression.
+                if (_analyzeForCollectionExpression && this.IsComplexElementInitializer(firstInit))
+                    return false;
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs:124]: needed to make this bool returning as we cannot use the "empty result means failure" approach for 'collection expressions'.  For collection expressions an empty result is fine and means "convert to `[]`".

File: src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs [hunk 10]
```diff
@@ -107,86 +149,157 @@ protected override void AddMatches(ArrayBuilder<Match<TStatementSyntax>> matches
                     continue;
                 }
 
-                if (extractedChild is TExpressionStatementSyntax expressionStatement)
+                if (extractedChild is TExpressionStatementSyntax expressionStatement &&
+                    TryProcessExpressionStatement(expressionStatement))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractUseCollectionInitializerAnalyzer.cs:153]: this large if/else was extracted into separate nested helper local functions.  

One for `x.Add(y)`, one for `foreach (var y in z) x.Add(y)` and one for `if (z) x.Add(y)`

File: src/Analyzers/Core/CodeFixes/UseCollectionInitializer/AbstractUseCollectionInitializerCodeFixProvider.cs [hunk 11]
```diff
@@ -97,15 +113,16 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
...
 
+            using var analyzer = GetAnalyzer();
+
             while (originalObjectCreationNodes.Count > 0)
             {
                 var (originalObjectCreation, useCollectionExpression) = originalObjectCreationNodes.Pop();
                 var objectCreation = currentRoot.GetCurrentNodes(originalObjectCreation).Single();
 
-                var matches = UseCollectionInitializerAnalyzer<TExpressionSyntax, TStatementSyntax, TObjectCreationExpressionSyntax, TMemberAccessExpressionSyntax, TInvocationExpressionSyntax, TExpressionStatementSyntax, TForeachStatementSyntax, TVariableDeclaratorSyntax>.Analyze(
-                    semanticModel, syntaxFacts, objectCreation, useCollectionExpression, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/CodeFixes/UseCollectionInitializer/AbstractUseCollectionInitializerCodeFixProvider.cs:106]: no need for consumption sides of the helper type ot have to have so much noise.  this was getting particularly awful as i needded to add 'IfStatementSyntax' here (and all the other places).

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 12]
```diff
@@ -1614,16 +1614,10 @@ public void GetPartsOfConditionalExpression(SyntaxNode node, out SyntaxNode cond
...
         {
             var commonForeach = (CommonForEachStatementSyntax)statement;
-            identifier = commonForeach is ForEachStatementSyntax { Identifier: var foreachIdentifier }
-                ? foreachIdentifier
-                : default;
-            expression = commonForeach.Expression;
-            statements = commonForeach.Statement is BlockSyntax block
-                ? block.Statements
-                : SpecializedCollections.SingletonEnumerable(commonForeach.Statement);
+            return commonForeach.Expression;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:1620]: example of how gnarly this was trying to be a syntaxfact (which i added in the previous PR).  was happy to remove this here.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb [hunk 13]
```diff
@@ -1813,15 +1813,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.LanguageService
             whenFalse = conditionalExpression.WhenFalse
         End Sub
 
-        Public Sub GetPartsOfForeachStatement(statement As SyntaxNode, ByRef identifier As SyntaxToken, ByRef expression As SyntaxNode, ByRef statements As IEnumerable(Of SyntaxNode)) Implements ISyntaxFacts.GetPartsOfForeachStatement
-            Dim foreachStatement = DirectCast(statement, ForEachStatementSyntax)
-            Dim declarator = TryCast(foreachStatement.ControlVariable, VariableDeclaratorSyntax)
-            identifier = If(declarator Is Nothing, Nothing, declarator.Names(0).Identifier)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:1819]: similar gnarlyness for VB.

Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:1819]: similar gnarlyness for VB.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs (hunk 12 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:1620]: example of how gnarly this was trying to be a syntaxfact (which i added in the previous PR).  was happy to remove this here.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb (hunk 13 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:1819]: similar gnarlyness for VB.

Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/SyntaxFacts/VisualBasicSyntaxFacts.vb:1819]: similar gnarlyness for VB.

**Comment by CyrusNajmabadi** [comment]
@genlu @Cosifne @akhera99 this is ready for review.  I'm sorry for al the churn, but i'm trying to have independent PRs that solve core issues, without something absolutely ginormous that rolls it all in.

However, it does mean that as new things are added, there's often reworking of the core logic to try to keep it clean/rational.

**Comment by CyrusNajmabadi** [comment]
NOte: after the next PR, things should be in a much nicer state.  So only 2 more to go with the deep refactorings :)

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal :)

**Comment by CyrusNajmabadi** [comment]
@akhera99 ptal :)

**Review by akhera99 — APPROVED** [review]

**Pull request merged**